[exporter/bmchelix] fix distinct datapoints merged into a single metric payload

NassimBtk opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description
This pull request addresses a bug in the `bmchelixexporter` component where distinct datapoints were being merged into a single metric payload. The changes include updates to the metric conversion logic, test cases, and the addition of a changelog entry.

### Bug Fix: Handling of Distinct Datapoints

* **Metric Conversion Logic Update**:
  - Replaced `createHelixMetric` with `createHelixMetrics` to generate individual `BMCHelixOMMetric` payloads for each datapoint instead of merging them into one. This ensures distinct datapoints are preserved. (`exporter/bmchelixexporter/internal/operationsmanagement/metrics_producer.go`)
  - Introduced `createSingleDatapointMetric` to handle the creation of metrics at the datapoint level. (`exporter/bmchelixexporter/internal/operationsmanagement/metrics_producer.go`)

* **Test Case Enhancements**:
  - Updated `TestProduceHelixPayload` to validate the creation of separate metrics for distinct datapoints. Added multiple datapoints in mock data and adjusted expected payload accordingly. (`exporter/bmchelixexporter/internal/operationsmanagement/metrics_producer_test.go`)

### Documentation and Changelog

* **Changelog Entry**:
  - Added a changelog entry to document the bug fix, specifying the change type as `bug_fix` and providing a brief description of the issue and resolution. (`.chloggen/bmchelixexporter_fix_datapoints_merge.yaml`)
<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
Fixes #40991

<!--Describe what testing was performed and which tests were added.-->
#### Testing

All entities are now automatically detected:
![image](https://github.com/user-attachments/assets/d5cee276-706f-4ec8-8865-f18880645e81)

Metric sent successfully
![image](https://github.com/user-attachments/assets/eb75ebc2-6fe0-40e9-a856-da0e74f4e77a)

NassimBtk committed:
[exporter/bmchelix] fix distinct datapoints merged into a single payload

- Replace `createHelixMetric` with `createHelixMetrics` for batch
processing
- Add support for multiple datapoints in `createHelixMetrics` logic
- Simplify `createSingleDatapointMetric` for individual datapoint
handling
- Update tests to validate multiple datapoints per metric
- Modify mock data generation to include multiple datapoints
- Adjust expected payloads in tests to reflect new structure

NassimBtk committed:
[exporter/bmchelix] fix distinct datapoints merged into a single payload

- Add changelog entry for bug fix in bmchelix component

NassimBtk committed:
[exporter/bmchelix] fix distinct datapoints merged into a single payload

- Fixed formatting in metrics_producer_test.go

NassimBtk committed:
[exporter/bmchelix] fix distinct datapoints merged into a single payload

- Cleaned up redundant blank line in `metrics_producer.go`.

bertysentry commented on exporter/bmchelixexporter/internal/operationsmanagement/metrics_producer.go:
```diff
@@ -106,12 +110,44 @@ func appendMetricWithParentEntity(helixMetrics []BMCHelixOMMetric, helixMetric B
+			helixMetrics = append(helixMetrics, *metricPayload)
+		}
+	default:
+		return nil, fmt.Errorf("unsupported metric type %s", metric.Type())
```
Maybe a "warning" is enough for this. It's normal to discard unsupported metrics.

NassimBtk replied:
Yes, `Errorf` formats the message using a format specifier and returns it as a value that implements the `error` interface (similar to an exception). In this context, the caller logs it as a warning, so this behavior should be good.

NassimBtk commented on exporter/bmchelixexporter/internal/operationsmanagement/metrics_producer.go:
```diff
@@ -61,13 +61,17 @@ func (mp *MetricsProducer) ProduceHelixPayload(metrics pmetric.Metrics) ([]BMCHe
-				helixMetrics = appendMetricWithParentEntity(helixMetrics, *newHelixMetric, containerParentEntities)
+				// Loop through the newly created metrics and append them to the helixMetrics slice
+				// while also creating parent entities for container metrics
+				for _, m := range newMetrics {
+					helixMetrics = appendMetricWithParentEntity(helixMetrics, m, containerParentEntities)
+				}
```
Yes, let's use `slices.Grow` to increase the slice's capacity if necessary.

NassimBtk committed:
[exporter/bmchelix] fix distinct datapoints merged into a single payload

- Improve performance by reducing slice reallocation overhead.
- Import `slices` package for slice manipulation.
- Pre-grow `helixMetrics` in `scopeMetric.Metrics` loop.
- Pre-grow `helixMetrics` for `MetricTypeSum` data points.
- Pre-grow `helixMetrics` for `MetricTypeGauge` data points.

NassimBtk committed:
Merge branch 'main' into bmchelix-fix-distinct-datapoints-merged-into-single-payload

NassimBtk committed:
[exporter/bmchelix] fix distinct datapoints merged into a single payload

- Add slice growth after filtering new metrics.

NassimBtk committed:
Merge branch 'bmchelix-fix-distinct-datapoints-merged-into-single-payload' of https://github.com/sentrysoftware/opentelemetry-collector-contrib.git into bmchelix-fix-distinct-datapoints-merged-into-single-payload
